### PR TITLE
Multiphase Monte Carlo Sampling

### PR DESCRIPTION
**Pull request by TolisChal**
This PR implements a Multiphase Monte Carlo Sampling (MMCS) algorithm that unifies sampling with distribution isotropization. The PR contains:

-- A C++ implementation of a single step of MMCS
-- A C++ class to update the Effective Sample Size for each new generated chain
-- A C++ implementation for the estimator of autocovariance in "Charles J. Geyer, Practical Markov Chain Monte Carlo, Statistical Science 1992."
-- New burn-in routines for Billiard Walk.
-- Two C++ examples to illustrate the new method.

This PR depends on https://github.com/GeomScale/volume_approximation/pull/142 and https://github.com/GeomScale/volume_approximation/pull/144

### CODE REVIEW

File: R-proj/test_ess.R [hunk 1]
```diff
@@ -0,0 +1,5 @@
+library(volesti)
```
Comment (papachristoumarios) [inline R-proj/test_ess.R]: That file should not be in the PR right?

File: examples/mmcs_method/random_hpoly_50_dim.cpp [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    NT max_s, s_cutoff = 3.0, L;
+    bool complete = false, request_rounding = true,
+         rounding_completed = false;
+    bool req_round_temp = request_rounding;
+
+    std::pair<Point, NT> InnerBall;
+    
+    MT S;
+
+    std::cout<<"target effective sample size = "<<Neff<<"\n"<<std::endl;
```
Comment (papachristoumarios) [inline examples/mmcs_method/random_hpoly_50_dim.cpp]: Please leave spaces between <<

File: include/diagnostics/ess_updater_utils.hpp [hunk 3]
```diff
@@ -0,0 +1,70 @@
+// VolEsti (volume computation and sampling library)
+// Copyright (c) 2021 Vissarion Fisikopoulos
+// Copyright (c) 2021 Apostolos Chalkis
+
+// Licensed under GNU LGPL.3, see LICENCE file
+
+
+#ifndef ESS_UPDATER_UTILS_HPP
```
Comment (papachristoumarios) [inline include/diagnostics/ess_updater_utils.hpp]: Add DIAGNOSTICS_ in front of the macro.

File: include/diagnostics/ess_updater_utils.hpp [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    using the Geyer's stable estimator given in 
+    Charles J. Geyer, Practical Markov Chain Monte Carlo, Statistical Science 1992.
+
+    * @tparam NT number type
+    * @tparam VT vector type
+    * @param samples the sequence of correlated samples
+    * @param auto_cov the autocovariance estimates
+*/
+template <typename NT, typename VT>
+void compute_autocovariance(VT const& samples, VT &auto_cov)
```
Comment (papachristoumarios) [inline include/diagnostics/ess_updater_utils.hpp]: Is it better the file having a different name, since we compute autocovariance?

**Review by papachristoumarios — CHANGES_REQUESTED** [review]
See my comments. Please also add tests.

In general, he PR is in very good condition in my opinion.

File: R-proj/test_ess.R (hunk 1 above)
Comment (TolisChal) [reply R-proj/test_ess.R]: Thank you for your feedback and comments. Yes I removed this file.

File: examples/mmcs_method/random_hpoly_50_dim.cpp (hunk 2 above)
Comment (TolisChal) [reply examples/mmcs_method/random_hpoly_50_dim.cpp]: Thanks. Done.

File: include/diagnostics/ess_updater_utils.hpp (hunk 3 above)
Comment (TolisChal) [reply include/diagnostics/ess_updater_utils.hpp]: Thanks. Done.

File: include/diagnostics/ess_updater_utils.hpp (hunk 4 above)
Comment (TolisChal) [reply include/diagnostics/ess_updater_utils.hpp]: Thanks. Done.

File: R-proj/src/copula.cpp [hunk 5]
```diff
@@ -7,6 +7,7 @@
 
 //Contributed and/or modified by Apostolos Chalkis, as part of Google Summer of Code 2018 program.
 
+
```
Comment (vissarion) [inline R-proj/src/copula.cpp]: many files have this empty line insertion, could you undo these changes?

File: include/diagnostics/ess_updater_autocovariance.hpp [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    auto_cov.setZero(N);
+
+    // compute normalized samples
+    VT normalized_samples(2 * N);
+    normalized_samples.setZero();
+    normalized_samples.head(N) = samples.array() - samples_mean;
+
+    NT variance = (normalized_samples.cwiseProduct(normalized_samples)).sum();
+    variance *= (1.0 / N);
+    variance += NT(1e-16)*(samples_mean*samples_mean);
```
Comment (vissarion) [inline include/diagnostics/ess_updater_autocovariance.hpp]: what is the purpose of this computation? why `1e-16`? Is this some magic number that simply works?

File: include/random_walks/uniform_accelerated_billiard_walk.hpp [hunk 7]
```diff
@@ -173,15 +248,15 @@ struct AcceleratedBilliardWalk
             T -= _lambda_prev;
             P.compute_reflection(_v, _p, _update_parameters);
 
-            while (it <= 100*n)
+            while (it <= 1000*n)
```
Comment (vissarion) [inline include/random_walks/uniform_accelerated_billiard_walk.hpp]: what is the reason behind this change? stability?

File: include/sampling/mmcs.hpp [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+<
+        typename Polytope,
+        typename RandomNumberGenerator,
+        typename MT,
+        typename Point,
+        typename WalkTypePolicy
+>
+void perform_mmcs_step(Polytope &P,
+                       RandomNumberGenerator &rng,
+                       const unsigned int &walk_length,
```
Comment (vissarion) [inline include/sampling/mmcs.hpp]: should this be `unsigned int const&` as below ?

File: include/sampling/mmcs.hpp [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+                       const unsigned int &target_ess,
+                       unsigned int const& max_num_samples,
+                       unsigned int const& window,
+                       unsigned int &Neff_sampled,
+                       unsigned int &total_samples,
+                       unsigned int const& num_rounding_steps,
+                       MT &TotalRandPoints,
+                       bool &complete,
```
Comment (vissarion) [inline include/sampling/mmcs.hpp]: I think is is clearer if this functions returns this `bool` and below you simply return true or false instead of setting this variable and return void.

File: include/sampling/parallel_mmcs.hpp [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+                                RandomNumberGenerator &rng,
+                                const unsigned int &walk_length,
+                                const unsigned int &target_ess,
+                                unsigned int const& max_num_samples,
+                                unsigned int const& window,
+                                unsigned int &Neff_sampled,
+                                unsigned int &total_samples,
+                                unsigned int const& num_rounding_steps,
+                                MT &TotalRandPoints,
+                                bool &complete,
```
Comment (vissarion) [inline include/sampling/parallel_mmcs.hpp]: same as above

**Review by vissarion — APPROVED** [review]
Great, thanks! I have only a few comments.

File: R-proj/src/copula.cpp (hunk 5 above)
Comment (TolisChal) [reply R-proj/src/copula.cpp]: done in `root/R-proj` folder

File: include/diagnostics/ess_updater_autocovariance.hpp (hunk 6 above)
Comment (TolisChal) [reply include/diagnostics/ess_updater_autocovariance.hpp]: Yes, it was added to avoid numerical overflow in case where `(normalized_samples.cwiseProduct(normalized_samples)).sum()` is zero.

Now I use `const NT eps = 1e-16;`

File: include/random_walks/uniform_accelerated_billiard_walk.hpp (hunk 7 above)
Comment (TolisChal) [reply include/random_walks/uniform_accelerated_billiard_walk.hpp]: Better performance in the initialization of the parameters when we have very skinny polytopes.

Now I use `unsigned int _rho = 1000*n;`

File: include/sampling/mmcs.hpp (hunk 8 above)
Comment (TolisChal) [reply include/sampling/mmcs.hpp]: done

File: include/sampling/parallel_mmcs.hpp (hunk 10 above)
Comment (TolisChal) [reply include/sampling/parallel_mmcs.hpp]: done

File: include/sampling/mmcs.hpp (hunk 9 above)
Comment (TolisChal) [reply include/sampling/mmcs.hpp]: done
